Docsp 25493 record codec topic (#327)
mongodb/docs-java: source/fundamentals/data-formats/document-data-format-record.txt

     - Specifies a BSON type to store when different from the record component
       type. Accepts the BSON type as the parameter.

.. note::

>    The driver supports these only on the record component.

Reviewer: also delete extra spacing below note

>    The driver supports these only on the record component.

Author: Good question - I was parroting what is on the API page, and am not quite sure. I'll ask in the docs-java channel for clarification.

     - Specifies a BSON type to store when different from the record component
       type. Accepts the BSON type as the parameter.

.. note::

>    The driver supports annotations for Java recrods, but only if you use them

Reviewer: ```suggestion
   The driver supports annotations for Java records, but only if you use them
```

>    when you define the component as shown in the example annotated record. You
>    cannot use the annotations inside the constructor.

Reviewer: ```suggestion
   when you define the component as shown in the following example record. You
   cannot use annotations inside the record constructor.
```

   .. output::
      :language: sh

      RecordTree[content=Ranginui, left=RecordTree[content=..., left=RecordTree[content=..., right=RecordTree[content=Ikatere...]]

.. _fundamentals-records-get-codec:

Retrieve the Record Codec
-------------------------

> When working with records, consider using the ``RecordCodecProvider`` to get the
> record codec for Java records. For more information about codecs and why you might want

Reviewer: S: Make the language more direct and avoid redundancy

```suggestion
You can use the ``RecordCodecProvider`` to retrieve the
record codec to work with Java records. For more information about codecs and why you might want
```

> record codec for Java records. For more information about codecs and why you might want
> to use them, see :ref:`Codecs <fundamentals-codecs>`. In summary, you use codecs

Reviewer: ```suggestion
record codec for Java records. To learn more about codecs and their usage, see the :ref:`Codecs guide<fundamentals-codecs>`. In summary, you use codecs
```

to encode and decode Java objects to and from corresponding BSON types while

> minimizing duplicate code.

Reviewer: S: move this sentence to the middle, and make it more specific to records

> .. tip::

Reviewer: S: I think this is better suited to be a note

>    When using the record codec, the annotations listed under
>    :ref:`Specify Component Conversion Using Annotations <fundamentals-records-annotations>`
>    are only supported on the record component.

Reviewer: S: avoid passive voice
```suggestion
   The driver supports the annotations listed under
   :ref:`Specify Component Conversion Using Annotations <fundamentals-records-annotations>`
   only on the record component.
```

Also, what does "record component" mean?

Author: I moved this to be in the section specifically about annotations because I think it fits much better there. Otherwise this information is scattered in a couple of places.

> Be aware that you do not create a record codec directly. Instead,
> use the ``RecordCodecProvider`` to get the record codec. See the 
> `RecordCodecProvider API page <{+api+}/apidocs/bson-record-codec/org/bson/codecs/record/RecordCodecProvider.html>`__
> for more information.

Reviewer: S: this language doesn't sound consistent with the general style of our docs

```suggestion
You can't create a record codec directly, but you can use the ``RecordCodecProvider`` to implement the record codec in your code. Learn more about the ``RecordCodecProvider`` in the `API documentation <{+api+}/apidocs/bson-record-codec/org/bson/codecs/record/RecordCodecProvider.html>`__.
```

The record codec supports nested generic records. These are used the same way as
other custom codecs as described on :ref:`Codecs <fundamentals-codecs>`. To see

> a complete example with custom codecs, see :ref:`Custom Codecs Example <codecs-custom-example>`.

Reviewer: C: This section seems out of place in this section, as it has nothing to do with retrieving the record codec. Is there another place you can move this content?

Author: Moved this to the beginning where we touch briefly on the record codec since I think it fits there much better. Good call out.

Author: So after moving this, I didn't like how it flowed. This seemed like it was something that didn't fit or should be called out. I went to put it in the What's New topic since I thought that was where it would be best to go. On doing that, I discovered it was already there, so I just deleted this section.

   .. output::
      :language: sh

      RecordTree[content=Ranginui, left=RecordTree[content=..., left=RecordTree[content=..., right=RecordTree[content=Ikatere...]]

.. _fundamentals-records-get-codec:

Retrieve the Record Codec

> -------------------------

Reviewer: Comment: I wonder if this whole section should be moved higher on the page.

Author: I don't think that is necessary. You only need to get the codec itself if you are creating a custom codec, which the rest of the topic does not get into. However, answering this comment, I realize that should be specified. I will add it to the text.

You can use the ``RecordCodecProvider`` to retrieve the
record codec. You should use this codec to encode and decode Java record objects

> to and from corresponding BSON types while minimizing duplicate code.

Reviewer: Q: how does using the RecordCodecProvider minimize duplicate code?

Author: This is inherent to the codec, so I feel it would be repetitive to add here. It is basically that you can just use the codec instead of needing to create BSON converters for each of your objects.

   .. output::
      :language: sh

      RecordTree[content=Ranginui, left=RecordTree[content=..., left=RecordTree[content=..., right=RecordTree[content=Ikatere...]]

.. _fundamentals-records-get-codec:

Retrieve the Record Codec
-------------------------

You can use the ``RecordCodecProvider`` to retrieve the

> record codec. You should use this when you want to customize the codec to encode

Reviewer: ```suggestion
record codec. You should use this interface when you want to customize the codec to encode
```

and decode Java record objects to and from corresponding BSON types while
minimizing duplicate code. To learn more about codecs and their usage,
see :ref:`Codecs <fundamentals-codecs>`.

You can't create a record codec directly, but you can use the

> ``RecordCodecProvider`` to implement the record codec in your code. Learn more
> about the ``RecordCodecProvider`` in the
> `API documentation <{+api+}/apidocs/bson-record-codec/org/bson/codecs/record/RecordCodecProvider.html>`__.

Reviewer: Convention

```suggestion
``RecordCodecProvider`` to implement the record codec in your code. To learn more
about the ``RecordCodecProvider``, see the
`API documentation <{+api+}/apidocs/bson-record-codec/org/bson/codecs/record/RecordCodecProvider.html>`__.
```
